packages/compass-saved-aggregations-queries/src/hooks/use-grid-filters.tsx

```diff
@@ -0,0 +1,263 @@
+});
+
+const searchInputStyles = css({
+  width: '300px',
+  marginRight: spacing[2],
+});
+
+const FilterSelect: React.FunctionComponent<{
+  options: string[];
+  onSelect: (value: string) => void;
+  value?: string;
+  placeholder?: string;
+}> = ({ options, onSelect, value, placeholder }) => {
+  const labelId = useId();
+  const controlId = useId();
+  const longestLabel = Math.max(
+    placeholder?.length || 0,
+    ...options.map((item) => item.length)
+  );
+  return (
+    <Select
+      disabled={options.length === 0}
+      id={controlId}
+      aria-labelledby={labelId}
+      allowDeselect={true}
+      placeholder={placeholder}
+      className={cx(
+        selectStyles,
+        css({ minWidth: `calc(${longestLabel}ch + ${spacing[6]}px)` })
+      )}
+      onChange={onSelect}
+      value={value}
+    >
+      {options.map((option) => (
+        <Option key={option} value={option}>
+          {option}
+        </Option>
+      ))}
+    </Select>
+  );
+};
+
+function useSearchFilter(): [React.ReactElement, string] {
+  const [search, setSearch] = useState('');
+  const searchControls = useMemo(() => {
+    return (
+      <TextInput
+        className={searchInputStyles}
+        aria-label="Search"
+        type="search"
+        placeholder="Search"
+        value={search}
+        onChange={(e) => {
+          setSearch(e.target.value);
+        }}
+        spellCheck={false}
+      />
+    );
+  }, [search]);
+
+  return [searchControls, search];
+}
+
+function useSelectFilter(items: Item[]): [React.ReactElement, SelectState] {
+  const [selectState, setSelectState] = useState<SelectState>({});
+  const [collections, setCollections] = useState<string[]>([]);
+
+  const databases = _.uniq(_.map(items, 'database'));
+
+  const selectDatabase = useMemo(() => {
+    return (database: string): void => {
+      setSelectState({
+        database: database ?? undefined,
+      });
+
+      const collections = _.uniq(
+        _.map(
+          _.filter(items, (x) => x.database === database),
+          'collection'
+        )
+      );
+      setCollections(collections);
+    };
+  }, [items]);
+
+  const selectCollection = useMemo(() => {
+    return (collection: string): void => {
+      setSelectState({
+        ...selectState,
+        collection: collection ?? undefined,
+      });
+    };
+  }, [selectState]);
+
+  const selectControls = useMemo(() => {
+    return (
+      <div className={selectContainer}>
+        <FilterSelect
+          options={databases}
+          placeholder={'All databases'}
+          onSelect={selectDatabase}
+          value={selectState.database}
+        />
+        <FilterSelect
+          options={collections}
+          placeholder={'All collections'}
+          onSelect={selectCollection}
+          value={selectState.collection}
+        />
+      </div>
+    );
+  }, [databases, selectDatabase, selectState, collections, selectCollection]);
+
+  return [selectControls, selectState];
+}
```

Comment:
When using react hooks you usually would want them to be focused on one thing, if you have to put an object in a useState, it usually is a good sign that you either want a use reducer (for complex cases) or just separate simple states, otherwise you are making it really hard to update in the future. Similar with useMemo here, because you are putting everything in one hook, you are loosing most of the benefits of memoization here. This case is very small, so don't feel obliged to change that, but as this is a small example, it's a good one to see the pattern (it's much harder to spot these issues on a bigger ones)